napari/utils/key_bindings.py

```diff
@@ -182,6 +186,284 @@ def normalize_key_combo(key_combo):
 UNDEFINED = object()
 
 
+def call_with_context(function, context):
+    """
+    call function `function` with the corresponding value taken from context
+
+    This is use in the action manager to pass the rights instances to the actions,
+    without having the need for them to take a **kwarg, and is mostly needed when
+    triggering actions via buttons, or to record.
+
+    If we went a declarative way to trigger action we cannot refer to instances
+    or objects that must be passed to the action, or at least this is
+    problematic.
+
+    We circumvent this by having a context (dictionary of str:instance) in
+    the action manager, and anything can tell the action manager "this is the
+    current instance a key". When an action is triggered; we inspect the
+    signature look at which instances it may need and pass this as parameters.
+    """
+    from inspect import signature
+
+    context_keys = [
+        k
+        for k, v in signature(function).parameters.items()
+        if v.kind not in (v.VAR_POSITIONAL, v.VAR_KEYWORD)
+    ]
+    ctx = {k: v for k, v in context.items() if k in context_keys}
+    return function(**ctx)
+
+
+@dataclass
+class Action:
+    command: Callable
+    description: str
+    keymappable: Any
```

Comment:
Even if something might sound Qt specific, it would be important in other contexts.

command does not currently make the distinction if the action is a `toggle`, or a `trigger`, for an action it might be unimportant, but for UI elements it might be important (if we move to using QToolBars)

Reply:
We could attach the shortcut to the action, as well, I mostly tried to keep the action pure data. I guess I could make an action an object that is evented when the shortcut changes. That would be fine with me as well.